cpp/src/lists/extract.cu

```diff
@@ -16,103 +16,136 @@
-    else
-      return index >= -length ? length + index + offset : out_of_bounds;
-  }
-};
+std::unique_ptr<cudf::column> make_index_child(column_view const& indices,
+                                               size_type ignore,  // Placeholder for number of rows.
+                                               rmm::cuda_stream_view stream)
+{
+  // New column, near identical to `indices`, except with null values replaced.
+  // `segmented_gather()` on a null index should produce a null row.
+  if (not indices.nullable()) { return std::make_unique<column>(indices, stream); }
+
+  auto const indices_device_view = column_device_view::create(indices);
+  auto const d_indices           = *indices_device_view;
+  auto const null_index          = std::numeric_limits<size_type>::max();
+  auto const null_replaced_iter_begin =
+    cudf::detail::make_null_replacement_iterator(d_indices, null_index);
+  auto index_child = cudf::make_numeric_column(
+    data_type{type_id::INT32}, indices.size(), mask_state::UNALLOCATED, stream);
+  thrust::copy_n(rmm::exec_policy(stream),
+                 null_replaced_iter_begin,
+                 indices.size(),
+                 index_child->mutable_view().begin<size_type>());
+  return index_child;
+}
+
+/**
+ * @brief Helper to construct a column of indices, for use with `segmented_gather()`.
+ *
+ * When indices are specified as a size_type, e.g. `7`,
+ * the column returned is: `{ 7, 7, 7, 7, 7 }`.
+ *
+ * The returned column can then be used to construct a lists column, for use
+ * with `segmented_gather()`.
+ */
+std::unique_ptr<cudf::column> make_index_child(size_type index,
+                                               size_type num_rows,
+                                               rmm::cuda_stream_view stream)
+{
+  auto index_child =  // [index, index, index, ..., index]
+    make_numeric_column(data_type{type_id::INT32}, num_rows, mask_state::UNALLOCATED, stream);
+  thrust::copy_n(rmm::exec_policy(stream),
+                 thrust::make_constant_iterator(size_type{index}),
+                 num_rows,
+                 index_child->mutable_view().begin<size_type>());
+  return index_child;
+}
+
+/**
+ * @brief Helper to construct offsets column for an index vector.
+ *
+ * Constructs the sequence: `{ 0, 1, 2, 3, ... num_lists + 1}`.
+ * This may be used to construct an "index-list" column, where each list row
+ * has a single element.
+ */
+std::unique_ptr<cudf::column> make_index_offsets(size_type num_lists, rmm::cuda_stream_view stream)
+{
+  auto index_offsets =  // [0, 1, 2, 3, ... num_lists + 1]
+    make_numeric_column(data_type{type_id::INT32}, num_lists + 1, mask_state::UNALLOCATED, stream);
+  thrust::copy_n(rmm::exec_policy(stream),
+                 cudf::detail::make_counting_transform_iterator(0, thrust::identity<size_type>{}),
+                 num_lists + 1,
+                 index_offsets->mutable_view().begin<size_type>());
```

Comment:
How about `sequence` instead?
```suggestion
  thrust::sequence(rmm::exec_policy(stream),
                   index_offsets->mutable_view().begin<size_type>(),
                   index_offsets->mutable_view().end<size_type>());
```
Actually, even better is to use the [`cudf::detail::sequence`](https://github.com/rapidsai/cudf/blob/6593339b2c388fd99fb8d091b1678898872ec054/cpp/include/cudf/detail/sequence.hpp#L48-L52) API to build the column for you.

Reply:
Fixed with `cudf::detail::sequence()`. Thanks for pointing this out.